utils/wasm-instrument/src/lib.rs

```diff
@@ -47,46 +50,173 @@ pub const GLOBAL_NAME_FLAGS: &str = "gear_flags";
+#[derive(Debug, PartialEq, Eq, derive_more::Display)]
+pub enum InstrumentationError {
+    /// Error occurred during injecting `gr_system_break` import.
+    #[display(fmt = "The WASM module already has `gr_system_break` import")]
+    SystemBreakImportAlreadyExists,
+    /// Error occurred during stack height instrumentation.
+    #[display(fmt = "Failed to inject stack height limits")]
+    StackLimitInjection,
+    /// Error occurred during injecting `gear_gas` global.
+    #[display(fmt = "The WASM module already has `gear_gas` global")]
+    GasGlobalAlreadyExists,
+    /// Error occurred during calculating the cost of the `gas_charge` function.
+    #[display(
+        fmt = "An overflow occurred while calculating the cost of the `gas_charge` function"
+    )]
+    CostCalculationOverflow,
+    /// Error occurred while trying to get the instruction cost.
+    #[display(fmt = "Failed to get instruction cost")]
+    InstructionCostNotFound,
+    /// Error occurred during injecting gas metering instructions.
+    ///
+    /// This might be due to program contained unsupported/non-deterministic instructions
+    /// (floats, memory grow, etc.).
+    #[display(fmt = "Failed to inject instructions for gas metrics: may be in case \
+        program contains unsupported instructions (floats, memory grow, etc.)")]
+    GasInjection,
+}
+
+/// This is an auxiliary builder that allows to instrument WASM module.
+pub struct InstrumentationBuilder<'a, R, GetRulesFn>
+where
+    R: Rules,
+    GetRulesFn: FnMut(&Module) -> R,
+{
+    /// name of module to import syscalls
+    module_name: &'a str,
+    /// configuration of stack_limiter
+    stack_limiter: Option<(u32, bool)>,
+    /// configuration of gas limiter
+    gas_limiter: Option<GetRulesFn>,
+}
 
-pub fn inject<R: Rules>(
+impl<'a, R, GetRulesFn> InstrumentationBuilder<'a, R, GetRulesFn>
+where
+    R: Rules,
+    GetRulesFn: FnMut(&Module) -> R,
+{
+    /// Creates a new [`InstrumentationBuilder`] with the given module name to import syscalls.
+    pub fn new(module_name: &'a str) -> Self {
+        Self {
+            module_name,
+            stack_limiter: None,
+            gas_limiter: None,
+        }
+    }
+
+    /// Whether to insert a stack limiter into WASM module.
+    pub fn with_stack_limiter(&mut self, stack_limit: u32, export_stack_height: bool) -> &mut Self {
+        self.stack_limiter = Some((stack_limit, export_stack_height));
+        self
+    }
+
+    /// Whether to insert a gas limiter into WASM module.
+    pub fn with_gas_limiter(&mut self, get_gas_rules: GetRulesFn) -> &mut Self {
+        self.gas_limiter = Some(get_gas_rules);
+        self
+    }
+
+    /// Performs instrumentation of a given WASM module depending
+    /// on the parameters with which the [`InstrumentationBuilder`] was created.
+    pub fn instrument(&mut self, module: Module) -> Result<Module, InstrumentationError> {
+        if !(self.stack_limiter.is_some() || self.gas_limiter.is_some()) {
```

Comment:
More readable
```suggestion
        if let (None, None) = (self.stack_limiter, self.gas_limiter) {
```

Reply:
ok, but need to add ref: `&self.gas_limiter` because type is not `Copy`
done in 342ae11dfc127005aab291a0ce01668a131195ee